### PR TITLE
Refactor Grid Traversal to use a simpler algorithm

### PR DESCRIPTION
**Pull request by jinlin270**
## Summary
Switch to use a very simple algorithm for lawnmower traversal by keeping track of the activated nodes in a 2D array during activation shape.

### Description
The algorithm works as follows: Suppose we have a grid like the one below, and we want to traverse the red activated region. During "activation", we will store the activated nodes into a 2D array by rows. If we want to traverse, we simply add one row at a time to the waypoints, and we turn and traverse in the opposite direction by reversing the order of the points in every other row. 
To traverse RIGHT, we flip every even row.
To traverse LEFT, we flip every odd row.
Similar logic applies for vertical traversal, by keeping track of a 2D array sorted by columns during activation.

IMG_3E96D7D0A3FA-1

Other changes:
* The algorithm keeps track of each node as a Node object, rather than a tuple of coordinates. This change is necessary during real life traversal so the longitude and latitude can be accessed.
* In practice, the algorithm only add the endpoints of each row or column, removing the unnecessary points in the middle to be more pid friendly.

Pros:
This algorithm gives a more intuitive understanding of the lawnmower traversal, and dramatically decreases the amount of code needed in grid.py. 
Cons:
The drawback to the reduced complexity is the increase in needed space by storing the 2D arrays of activated nodes. However, it shouldn't be too much of an issue considering we need to keep track of the entire grid  of nodes regardless, so the space complexity remains the same. It can also be mitigated if we choose to store only the endpoints of each row or column during activation
## Test Plan

python3 -m tests.functionality_tests.test_activation_shapes
Activation type: circle, rectangle, triangle, line
Traversal Direction: LEFT, RIGHT, DOWN, UP

## Other

TODO:
1. Implement turning algorithm to yield a list of Nodes. 
2. Retire prior code in grid.py if we do move forward with this algorithm
3. If want to further optimize, change the activation shapes so instead of storing the entire row or columns, store only the end points

Credits to Ahan for the code in test_activation_shapes.py!

### CODE REVIEW

File: engine/grid.py [hunk 1]
```diff
@@ -248,31 +282,38 @@ def area(p1, p2, p3):
         # is same as A
         return A == (A1 + A2 + A3)
 
-    def activate_line(self, coordinate, n, is_horizontal):
+    def activate_line(self, coordinate, n, is_vertical):
         """
         Activate a single horizontal line of length n starting at row, col.
```
Comment (AC408) [inline engine/grid.py:287]: update specs

File: engine/grid.py [hunk 2]
```diff
@@ -248,31 +282,38 @@ def area(p1, p2, p3):
...
+        if not is_vertical:
             for i in range(n):
-                self.nodes[col+i][row].is_active = True
+                self.nodes[col + i][row].is_active = True
+                self.horizontal_waypoint_list.append([self.nodes[col + i, row]])
         else:
             for i in range(n):
-                self.nodes[col][row+i].is_active = True
+                self.nodes[col][row + i].is_active = True
+                self.vertical_waypoint_list.append([self.nodes[col, row + i]])
```
Comment (AC408) [inline engine/grid.py:302]: for loop, then conditional expression for faster computation

File: engine/grid.py [hunk 3]
```diff
@@ -248,31 +282,38 @@ def area(p1, p2, p3):
...
 
     def activate_triangle(self, p1, p2, p3):
         """
         Activates all the nodes in a triangle based on three vertices p1, p2, p3.
         """
+        self.horizontal_waypoint_list = [[] for _ in range(self.num_cols)]
```
Comment (AC408) [inline engine/grid.py:308]: can we initialize this in the class constructor and only need to change it for activate_rectangle?

File: engine/grid.py [hunk 4]
```diff
@@ -211,26 +222,49 @@ def activate_rectangle(self, row, col, row_limit, col_limit):
         Activates all the nodes in a rectangle based on row start/end, col
         start/end.
         """
+        self.horizontal_waypoint_list = [[] for _ in range(col, col_limit)]
```
Comment (AC408) [inline engine/grid.py:225]: is col_limit - col = num_col?

File: engine/grid.py [hunk 5]
```diff
@@ -211,26 +222,49 @@ def activate_rectangle(self, row, col, row_limit, col_limit):
...
+            self.vertical_waypoint_list.append(col_list)
         return self.nodes
 
     def activate_circle(self, circle_center_row, circle_center_col, circle_radius):
         """
         Activates all the nodes in a circle based on center and radius.
         """
+        self.horizontal_waypoint_list = [[] for _ in range(self.num_cols)]
         for x in range(self.num_rows):
+            col = []
```
Comment (AC408) [inline engine/grid.py:242]: nit, naming invonsistencies

File: engine/grid.py [hunk 6]
```diff
@@ -600,6 +699,58 @@ def add_guided_waypoints(self, is_vertical):
...
+                if flip:
+                    self.waypoints += [
+                        self.vertical_waypoint_list[row][-1],
+                        self.vertical_waypoint_list[row][0],
+                    ]
+                else:
+                    self.waypoints += [
+                        self.vertical_waypoint_list[row][0],
+                        self.vertical_waypoint_list[row][-1],
+                    ]
```
Comment (AC408) [inline engine/grid.py:728]: funny way to simplify:
```
tmp = self.vertical_waypoint_list[row] 
self.waypoints += [tmp[-flip], tmp[-(not flip)]]
```
not sure if its a good change

File: engine/grid.py [hunk 7]
```diff
@@ -600,6 +699,58 @@ def add_guided_waypoints(self, is_vertical):
...
+                if flip:
+                    self.waypoints += [
+                        self.horizontal_waypoint_list[col][-1],
+                        self.horizontal_waypoint_list[col][0],
+                    ]
+                else:
+                    self.waypoints += [
+                        self.horizontal_waypoint_list[col][0],
+                        self.horizontal_waypoint_list[col][-1],
+                    ]
```
Comment (AC408) [inline engine/grid.py:750]: can be made into helper function

File: engine/grid.py [hunk 8]
```diff
@@ -600,6 +699,58 @@ def add_guided_waypoints(self, is_vertical):
...
+                flip = not flip
```
Comment (AC408) [inline engine/grid.py:751]: can be put outside if else

File: tests/functionality_tests/test_activation_shapes.py [hunk 9]
```diff
@@ -18,13 +17,26 @@ def main():
         print("flip rows, cols for traversal")
 
     activation_type = input("Activation type: ")
-    is_vertical = "True" == input("Vertical traversal: ")
-    if is_vertical:
+    direction = input("Traversal Direction: ")
+    if direction.isnumeric():
+        numberToDirection = {"0": "RIGHT", "1": "UP", "2": "LEFT", "3": "DOWN"}
```
Comment (AC408) [inline tests/functionality_tests/test_activation_shapes.py:22]: im ngl, i dont know if this is necessary to add; i dont know who would associate 0 to right, 1 to up, etc

File: tests/functionality_tests/test_activation_shapes.py [hunk 10]
```diff
@@ -18,13 +17,26 @@ def main():
...
+        direction = numberToDirection.get(direction)
+    else:
+        direction = direction.upper()
+
+    print(direction)
```
Comment (AC408) [inline tests/functionality_tests/test_activation_shapes.py:27]: remove print

File: tests/functionality_tests/test_activation_shapes.py [hunk 11]
```diff
@@ -18,13 +17,26 @@ def main():
...
+    if direction == "UP":
         grid.direction = grid.Direction.UP
+    elif direction == "DOWN":
+        grid.direction = grid.Direction.DOWN
+    elif direction == "RIGHT":
+        grid.direction = grid.Direction.RIGHT
+    else:
+        grid.direction = grid.Direction.LEFT
```
Comment (AC408) [inline tests/functionality_tests/test_activation_shapes.py:35]: i think else condition should be elif and add raise error for else

**Review by AC408 — CHANGES_REQUESTED** [review]
minor comments; can you resolve merge conflicts before making pr?

File: engine/grid.py (hunk 3 above)
Comment (jinlin270) [reply engine/grid.py:308]: This is also needed for shapes like circle or any slanted shapes we may have

File: engine/grid.py (hunk 4 above)
Comment (jinlin270) [reply engine/grid.py:225]: I think? I was following the logic in the original code

File: engine/grid.py (hunk 1 above)
Comment (jinlin270) [reply engine/grid.py:287]: Done

File: tests/functionality_tests/test_activation_shapes.py (hunk 11 above)
Comment (jinlin270) [reply tests/functionality_tests/test_activation_shapes.py:35]: Done

File: tests/functionality_tests/test_activation_shapes.py (hunk 9 above)
Comment (jinlin270) [reply tests/functionality_tests/test_activation_shapes.py:22]: It doesn't hurt to keep it? Also credits to Ahan! I test_activation_shapes.py from his PR and it's great.

File: engine/grid.py (hunk 6 above)
Comment (jinlin270) [reply engine/grid.py:728]: Smart! not as readable though... I think we can afford to keep it without the optimization

File: engine/grid.py (hunk 3 above)
Comment (AC408) [reply engine/grid.py:308]: yeah, but activate_circle uses that exact same initialization

File: tests/functionality_tests/test_activation_shapes.py (hunk 9 above)
Comment (AC408) [reply tests/functionality_tests/test_activation_shapes.py:22]: ig not, i dont know if it'll crowd the code space

File: engine/grid.py (hunk 3 above)
Comment (jinlin270) [reply engine/grid.py:308]: Good point

File: engine/grid.py [hunk 12]
```diff
@@ -248,31 +282,38 @@ def area(p1, p2, p3):
...
         """
+        self.horizontal_waypoint_list = [[] for _ in range(self.num_cols)]
         for x in range(self.num_rows):
+            col = []
             for y in range(self.num_cols):
                 if self.is_inside_triangle(p1, p2, p3, (x, y)):
                     self.nodes[x, y].is_active = True
+                    self.horizontal_waypoint_list[y].append(self.nodes[x, y])
+                    col.append(self.nodes[x, y])
+            self.vertical_waypoint_list.append(col)
```
[the remark this answers is not in the document]
Comment (jinlin270) [reply engine/grid.py:316]: There doesn't seem like this is an issue as long as is_inside_triangle is correct (which I believe it is with the condition that the points are on the positive x,y coordinate. I added the pre-condition for now, and we can make it a future todo to make it work for the entire x-y plane if needed. 
I am not sure if this is what you are talking about?

File: engine/grid.py (hunk 7 above)
Comment (jinlin270) [reply engine/grid.py:750]: thank you! Done

**Review by ahanbm — CHANGES_REQUESTED** [review]
Couple small changes: Can you remove the comments above simple_get_guided_waypoints (or make it into a separate method if you think the previous method might still have some utility). Can you also add a short bit of documentation to the simple_get_guided_waypoints method.

**Review by AC408 — APPROVED** [review]

**Review by ahanbm — APPROVED** [review]
Looks good!

**Pull request merged**